### ISSUE TITLE
Suggestion for rule about function try blocks of constructors/destructors

### ISSUE THREAD

**Issue by tcorbat** [isocpp/CppCoreGuidelines#1196]
I have a suggestion for an additional rule regarding function try blocks for constructors/destructors. As [except.handling]/14 [1] states that if control flow reaches the end of a handler of a constructor’s/destructor’s function try block, the handled exception is rethrown. To me (and I think for others it might be too) this case is surprising and therefore I suggest to always state this behavior explicitly (at least in the case mentioned), i.e. insert a `throw;` or `return` statement at the end of all handlers of constructor’s/destructor’s function try block. I'm curious whether others consider this as an issue too.

[1] http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4741.pdf

Possible rule:

### Reason:
If controlflow reaches the end of a constructor's/destructor’s handler of a function try block the handled exception will be implicitly rethrown. The behavior is different in any other catch handler. Therefore, the desired behavior should be stated explicitly in this case.

### Example, bad:
```
struct Resource {
    Resource() noexcept(false) try
    {
        allocationThatMightThrow();
    }
    catch (const AllocationException & e)
    {
        doSomeCleanup(e);
    }
};
```

### Example:
```
struct Resource {
    Resource() noexcept(false) try
    {
        allocationThatMightThrow();
    }
    catch (const AllocationException& e)
    {
        doSomeCleanup(e);
        throw;
    }
};
```

### Enforcement:
Flag catch handlers of a constructor’s function try block without explicit `throw` statement. Flag catch handlers of a destructor’s function try block without explicit `return` or `throw` statement.

**Comment by jwakely** [comment]
The example isn't a good use of a function try block. I think "don't use function try blocks" is a better guideline that applies more often.

**Comment by beinhaerter** [comment]
I disagree with the proposal as written now because **except.handle 13** says
> If a return statement (9.6.3) appears in a handler of the function-try-block of a constructor, the program is ill-formed.

This contradicts your rule **Explicitly return** or throw **at end of constructor’s**/destructor’s **function try block**.

**Comment by jwakely** [comment]
> Suggesting to avoid function try blocks (at least for constructors/destructors)

Avoid them everywhere.

**Comment by cubbimew** [comment]
> Avoid them everywhere.

I was going to say "what about converting and rethrowing exceptions", but then I never had to do that myself.. So I just ran Debian Code Search for [\)\s*try\s*:](https://codesearch.debian.net/search?q=%5C%29%5Cs*try%5Cs*%3A) (not perfect since it won't catch ones with line breaks) and it found 243 uses in [30 packages](https://codesearch.debian.net/results/85ea2c10be868390/packages.json), Almost all of them were compiler/IDE/other C++ parser test suites or that regex latching on to something that isn't C++.

I only found three uses in the entire Debian:

1. [MongoDB](https://github.com/mongodb/mongo/blob/master/src/mongo/stdx/thread.h#L80-L84) where they convert an exception in a constructor to std::terminate (because they don't allow exceptions)
2. Something called [performous](https://github.com/performous/performous/blob/ce14aa71f229ae17554b717e83158f0b07825748/game/songparser.cc#L44) where they used it [to convert standard exceptions](https://github.com/performous/performous/blob/ce14aa71f229ae17554b717e83158f0b07825748/game/songparser.cc#L95-L101) from member initializers to their own SongParserException.... until earlier this year when they [removed the function-try-block.](https://github.com/performous/performous/commit/0006ea617f1831efc1e8d567d7042ebd1f89faa0)
3. Something called [nordugrid-arc](https://sources.debian.org/src/nordugrid-arc/5.4.2-3/src/hed/dmc/srm/srmclient/SRMURL.cpp/?hl=11#L11) which doesn't seem to realize the exception is always rethrown and attempts to create an invalid object by executing `valid = false;` in the catch clause.

+1 for "Avoid them everywhere".

**Comment by arvidn** [comment]
Although I recognise most functions shouldn't have function-try blocks, the suggestion to *avoid* it, is it because:

1. the need for a function-try block indicates some design flaw that should be fixed instead.

or

2. the syntax is unfamiliar to people, so the more verbose version should be used instead.

If (2), I suppose this would apply to putting try-catch on a `for`-, `while`- and `if`-block as well. I do use try-blocks as `for`-loop bodies sometimes, if they handle events that are allowed to fail independently, without preventing all subsequent events from being executed.

I have a few example of cases where function try-blocks are convenient. For instance (because they save an indentation level):

1. A function is best-effort and all errors are just logged, not reported up the stack
2. Some specific kinds of errors are recorded by some other mean without interrupting the program flow of the caller. [example1](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/alert_manager.hpp#L75),  [example2](https://github.com/arvidn/libtorrent/blob/master/src/torrent.cpp#L2477). Perhaps the most compelling case for this is boost.asio handlers. The majority of errors that happens inside them should not be propagated up into the asio executor, but handled some other way. For instance by having a handler allocator that catches exceptions and lets the underlying object handle the error, ([example](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/aux_/allocating_handler.hpp#L97)).
3. A function whose only purpose is to handle the current exception (i.e. re-throw and implement the catch handlers). [example](https://github.com/arvidn/libtorrent/blob/master/src/torrent.cpp#L1043)
4. The `main()` function of a utility where any error propagated to that level should print an error and return a non-zero status code. [example](https://github.com/arvidn/libtorrent/blob/master/examples/bt-get.cpp#L43)

@cubbimew I suppose my code didn't show up in your code search.

**Comment by ChrisGuzak** [comment]
Windows uses function try on ABI boundaries with the main reason being concision. I like @arvidn points and think the main() example is illustrative and I have written that myself many times. if guidelines say never I need to learn more about what I'm getting wrong.

**Comment by gdr-at-ms** [comment]
We will clarify the existing rules about function-try block for constructors/destructor.

**Comment by jwakely** [comment]
@arvidn 

> If (2), I suppose this would apply to putting try-catch on a for-, while- and if-block as well. 

No, not at all. A `try-catch` block is a statement, so I've never met anybody who is surprised that you can use it as the body of a `for` or `while` loop. It's no different to a single statement without braces:

    while (cond)
      ++i; /* a single statement

    while (cond)
      { /* a block statement */; }

    while (cond)
      try { } catch (...) { };

But a function try block is unique syntax, allowing `try` to appear outside a function body. You can't do that with anything else:

    void func() while (cond) { }

    void func(int i) ++i; { }

It's a completely false analogy to say that if people aren't familiar with function try blocks they aren't familiar with other perfectly normal uses of them.

**Comment by arvidn** [comment]
> It's a completely false analogy to say that if people aren't familiar with function try blocks they aren't familiar with other perfectly normal uses of them.

It is my personal experience that it's common to learn about these uses of try-catch blocks together, and colleagues being surprised of both. But obviously I can't claim this is the case universally.

**Comment by BjarneStroustrup** [comment]
My impression is that function try blocks are little known because many, probably most, textbooks don't mention them.
